### PR TITLE
Smart Factors for arbitrary cameras

### PR DESCRIPTION
**Pull request by lucacarlone**
Smart Factors that can deal with arbitrary cameras. 

@dellaert : I suggest merging the PR "feature/rollingShutterSmartFactors" first. If possible, can you also add my student Marcus Abate mabate@mit.edu to GTSAM and add as reviewer here?

### CODE REVIEW

**Comment by dellaert** [comment]
I will add mailto:mabate@mit.edu. I would rather not have @varunagrawal soend time reviewing this as he should be focusing on ICRA.

**Comment by lucacarlone** [comment]
> I will add mailto:[mabate@mit.edu](mailto:mabate@mit.edu). I would rather not have @varunagrawal soend time reviewing this as he should be focusing on ICRA.

perfect!

**Comment by varunagrawal** [comment]
Yeah I wasn't going to look at this until after September 15th.

**Comment by lucacarlone** [comment]
> Yeah I wasn't going to look at this until after September 15th.

up to you and @dellaert : not urgent, but ready to go.

**Comment by lucacarlone** [comment]
@dellaert : I finished the merge and went smoothly. This branch is ready when you have time.

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:45]: Format entire file with Google style?

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:112]: Const &? Maybe check other factors as well

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:123]: I am a bit worried about the design. Think about 10,000 time steps of a 6 camera rig. Will we have 60,000 calibrations? And 60,000 body_P_sensors?

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:138]: Const &

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:151]: Return a const&. No inline.

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:156]: Same

File: gtsam/slam/SmartProjectionFactorP.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionFactorP.h:161]: Same

**Review by dellaert — COMMENTED** [review]
I have two major concerns with this PR
- the storage required for multi-camera rigs . It seems to me that a fixed camera rig should be handled in a different way. 
- the copy/paste nature of the PR. What refactoring could be done to cut back on this? I had the same question with previous PR, and  it now becomes heavy.

**Comment by lucacarlone** [comment]
> I have two major concerns with this PR
> 
> * the storage required for multi-camera rigs . It seems to me that a fixed camera rig should be handled in a different way.
> * the copy/paste nature of the PR. What refactoring could be done to cut back on this? I had the same question with previous PR, and  it now becomes heavy.

both understandable @dellaert - for the concern about storage: easy to convert storage from vectors to maps from calibration "key" to objects and reduce storage. I'm happy to do that if you are onboard.

The issue with copy-paste is more subtle: the factor is developed to follow the same structure as the rolling shutter factors, but the functionality is quite different. Maybe the way to go is to create a base class for both that inherits from SmartProjectionFactor?

**Comment by dellaert** [comment]
@lucacarlone hold off on making major changes. I'm creating a PR on your branch, coming shortly, where I fixed docs and formatting as I bring myself up to speed again.

**Comment by lucacarlone** [comment]
@dellaert : I applied the changes we discussed. I think this is ready to go (after all CI is done). Please double-check the formatting for the factors if you can: I still don't trust my eclipse style settings :-(

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionRigFactor.h:103]: I don't understand why we have this constructor, but if you must, delegate to constructor above using { camera}

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionRigFactor.h:215]: make temp variable in loop for cameraRig_[cameraIds_[i]]

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (dellaert) [inline gtsam/slam/SmartProjectionRigFactor.h:245]: Weird that another set of cameras is passed in here. I though we just had poses and the camera rig, OR cameras computed from them. Sanity-check?

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:107]: same comment, user can just call above constructor with {camera}

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:96]: same comment on throwing error

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:61]: same comment on shared pointers

**Review by dellaert — APPROVED** [review]
Some more comments, but I won't review again. Please address and merge at will :-)

Yay !!!!

File: gtsam/slam/ReadMe.md
Comment (lucacarlone) [reply gtsam/slam/ReadMe.md:41]: done!

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h:43]: solved

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (marcusabate) [inline gtsam/slam/SmartProjectionRigFactor.h:126]: Should be `const MEASUREMENT& measured` so that other camera types can be used. With this it works with #861

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (marcusabate) [inline gtsam/slam/SmartProjectionRigFactor.h:146]: `const MEASUREMENTS& measurements` here as well

File: gtsam/slam/SmartProjectionFactorP.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionFactorP.h]: done!

File: gtsam/slam/SmartProjectionFactorP.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionFactorP.h]: solved as discussed

File: gtsam/slam/SmartProjectionFactorP.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionFactorP.h]: done

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h:67]: @dellaert : just for me to learn more: where do you expect savings from a shared ptr here? (which copy constructors are problematic in this file?)

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h:267]: the cameras here include the body poses (from the values) and the sensor-to-body poses (fixed, from the rig). The function "cameras" is in charge of assembling them to recover the sensor poses in the global frame. Please unresolve if this is still of concern.

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:116]: this does not work. Maybe because a cameraSet is more than just a set of cameras?

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h]: I'll apply this in #861. can you add this comment there?

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h:168]: actually I accepted this comment

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h]: scratch that: the "Measurement" infrastructure is only introduced in #861 , so this does not compile here. I'll need to apply your suggestions after I merge.

**Comment by lucacarlone** [comment]
@dellaert : FYI: I plan to merge this to develop as soon as the status check passes: keeping this in sync with develop and my other branches has been time-consuming and we can address the 2 concerns above (shared ptrs for camera rig and constructor with single camera) in a separate PR.

**Comment by dellaert** [comment]
> @dellaert : FYI: I plan to merge this to develop as soon as the status check passes: keeping this in sync with develop and my other branches has been time-consuming and we can address the 2 concerns above (shared ptrs for camera rig and constructor with single camera) in a separate PR.

Sounds like a plan. I seem to have missed your replies on my shared pointer comments. Did you end up using the shared pointers?

**Comment by lucacarlone** [comment]
> > @dellaert : FYI: I plan to merge this to develop as soon as the status check passes: keeping this in sync with develop and my other branches has been time-consuming and we can address the 2 concerns above (shared ptrs for camera rig and constructor with single camera) in a separate PR.
> 
> Sounds like a plan. I seem to have missed your replies on my shared pointer comments. Did you end up using the shared pointers?

thanks! I did not, since I was not sure if you wanted a shared pointer to CameraSet, or a vector of shared pointers to cameras. Also, I was not sure about which copy constructors may benefit from this (I do not see many in the code, and I was curious about which ones you were looking at).

File: gtsam/slam/SmartProjectionRigFactor.h
Comment (dellaert) [reply gtsam/slam/SmartProjectionRigFactor.h:67]: Ok, assume you have 10K factors, each taken with the *same* 12-camera rig, then you will 120000 camera objects, each with a copy of the camera, b ration, etc. 10K copies of all 12 cameras.

come to think of it, I would remove the second constructor (rethorical: why is it there?), and just pass *one* shared pointer to a Cameras object in the remaining constructor, so you just store one shared pointer per factor. Storage overhead is now negligible.

Comment (lucacarlone) [reply gtsam/slam/SmartProjectionRigFactor.h:67]: understood. will do before merging then.

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:62]: done

**Pull request merged**
